Centralize custom dimensions sync calls

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #7815

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/custom-dimensions.test.js:
```diff
@@ -276,6 +276,92 @@ describe( 'modules/analytics-4 custom-dimensions', () => {
+					.dispatch( MODULES_ANALYTICS_4 )
+					.scheduleSyncAvailableCustomDimensions();
+
+				const WAIT_AFTER_DISPATCH = 2500;
+
+				// Wait for more than two seconds as the sync is run after two seconds.
+				await new Promise( ( resolve ) => {
+					setTimeout( resolve, WAIT_AFTER_DISPATCH );
+				} );
```
We should use jest fake timers and related functions to advance the time or run all timers rather than doing a real wait for this long in the tests.

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/custom-dimensions.test.js:
```diff
@@ -276,6 +276,92 @@ describe( 'modules/analytics-4 custom-dimensions', () => {
+					clearInterval( intervalID );
+				}, DISPATCH_COUNT * DISPATCH_INTERVAL );
+
+				// Wait for 5 seconds as the sync is run two seconds after
+				// the last dispatch.
+				await new Promise( ( resolve ) => {
+					setTimeout(
+						resolve,
+						DISPATCH_COUNT * DISPATCH_INTERVAL + WAIT_AFTER_DISPATCH
+					);
+				} );
```
Same here

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js:
```diff
@@ -303,6 +306,90 @@ const baseSelectors = {
+		( select ) => ( state, customDimension ) => {
+			invariant( customDimension, 'customDimension is required.' );
+
+			const propertyID = select( MODULES_ANALYTICS_4 ).getPropertyID();
+
+			if ( ! propertyID ) {
+				return undefined;
+			}
+
+			const property =
+				select( MODULES_ANALYTICS_4 ).getProperty( propertyID );
```
These selectors wouldn't work in a view-only context. The sync shouldn't be attempted in that context anyways so I guess let's make sure that's still the case 😄

nfmohit replied:
Understood. For a view-only user, since the property info will be undefined, the custom dimension will be immediately considered to be "gathering data", and hence the report request to check the gathering data state will not be made. So, it looks like we're safe in this context.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js:
```diff
@@ -303,6 +306,90 @@ const baseSelectors = {
+			);
+
+			return customDimensions.reduce( ( errors, customDimension ) => {
+				const reportArgs =
+					select(
+						MODULES_ANALYTICS_4
+					).getDataAvailabilityReportOptions( customDimension );
+
+				if ( ! reportArgs ) {
+					return errors;
+				}
+
+				const error = select( MODULES_ANALYTICS_4 ).getErrorForSelector(
```
Please get the selectors outside of the per-dimension loop rather than part of every iteration (i.e. outside of `.reduce`).

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/custom-dimensions.js:
```diff
@@ -195,6 +199,70 @@ const baseActions = {
+export const baseControls = {
+	[ SCHEDULE_SYNC_AVAILABLE_CUSTOM_DIMENSIONS ]: createRegistryControl(
+		( { select, dispatch } ) =>
+			() => {
```
This function could be a bit cleaner by destructuring action and selector functions out at the top.

nfmohit replied:
Updated, thanks!

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/custom-dimensions.js:
```diff
@@ -195,6 +199,70 @@ const baseActions = {
+						MODULES_ANALYTICS_4
+					).fetchSyncAvailableCustomDimensions();
+
+					dispatch( MODULES_ANALYTICS_4 ).setSyncTimeoutID(
+						undefined
+					);
```
I'm not sure this needs to be unset aside from the above where it calls `clearTimeout`. Since the timeout is really a side-effect specific to the control, the reference to it doesn't even need to be in the store I don't think (it could just be a private variable in scope here).

nfmohit replied:
> Since the timeout is really a side-effect specific to the control, the reference to it doesn't even need to be in the store I don't think (it could just be a private variable in scope here).

Okay, I tried doing this with a scoped variable and also using lodash's `debounce` function. It seems the timeout ID isn't found across multiple dispatches, and the fetch is run as many times as the action is dispatched, just with the 2-second delay. Any ideas?

> I'm not sure this needs to be unset aside from the above where it calls `clearTimeout`.

Since we're using the existing pattern of keeping the timeout ID in the state (for now, unless we find a better solution), I think it might be worth unsettling this here just to make sure `clearTimeout` isn't unnecessarily called on the next dispatch. WDYT?

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/custom-dimensions.js:
```diff
@@ -326,12 +400,27 @@ const baseSelectors = {
+				select(
+					MODULES_ANALYTICS_4
+				).isFetchingSyncAvailableCustomDimensions() ||
+				!! state?.syncTimeoutID
```
I'm not sure we need to reference the timeout ID here, do we? I think isSyncing can probably remain as it was 🤔

nfmohit replied:
I've removed the selector entirely (and updated its usage) as it just uses `isFetchingSyncAvailableCustomDimensions` under the hood. Thanks!

nfmohit replied:
I actually added it because it causes a 2 second (or more for multiple tiles) flicker between the loading state and the error state when the report request fails due to an invalid custom dimensions error and needs to re-sync. Thoughts?

aaemnnosttv commented on assets/js/modules/analytics-4/utils/withCustomDimensions.js:
```diff
@@ -270,29 +308,31 @@ export default function withCustomDimensions( options = {} ) {
+				// triggered multiple times.
+				Promise.all(
+					invalidCustomDimensionReportOptions.map( ( args ) => {
+						return clearError( 'getReport', [ args ] );
```
The `return` here is unnecessary, correct?

nfmohit replied:
Even though the `return` is not necessary here, ESLint complains about the arrow function not having a return statement. We're also using a similar pattern in `create-snapshot-store.js`, e.g.

https://github.com/google/site-kit-wp/blob/14d70dea611e7de142ce3f6c122f9be15bbcae5a/assets/js/googlesitekit/data/create-snapshot-store.js#L187-L214

Thoughts?

aaemnnosttv commented on assets/js/modules/analytics-4/utils/withCustomDimensions.js:
```diff
@@ -270,29 +308,31 @@ export default function withCustomDimensions( options = {} ) {
+					invalidCustomDimensionReportOptions.map( ( args ) => {
+						return clearError( 'getReport', [ args ] );
+					} )
+				).finally( () => {
+					// Sync available custom dimensions.
+					scheduleSyncAvailableCustomDimensions();
```
Since the above `Promise.all` is only clearing errors and can't throw, let's update this to use `await` and then call this afterwards without using `finally` or `then`.

nfmohit replied:
Updated, thanks! I had only added it to avoid using `async-await` inside a `useEffect` 😅

aaemnnosttv commented on includes/Modules/Analytics_4/Custom_Dimensions_Data_Available.php:
```diff
@@ -111,9 +111,14 @@ public function set_data_available( $custom_dimension ) {
 	 */
-	public function reset_data_available() {
-		foreach ( self::CUSTOM_DIMENSION_SLUGS as $custom_dimension ) {
+	public function reset_data_available(
```
Please add/update the test coverage for this method accordingly, thanks!

nfmohit replied:
Added, thank you!

aaemnnosttv commented on assets/js/modules/analytics-4/utils/withCustomDimensions.js:
```diff
@@ -228,11 +220,57 @@ export default function withCustomDimensions( options = {} ) {
+								)
+							)
+							.map( ( dimension ) =>
+								getDataAvailabilityReportOptions( dimension )
```
If we have the error object, we can reverse lookup its report args via `getMetaDataForError` by the error itself, so maybe this selector isn't necessary?

nfmohit replied:
That is correct, however, the `getDataAvailabilityReportErrors` selector uses the `getDataAvailabilityReportOptions` selector under the hood, so we aren't able to remove it. Is it okay to keep using `getDataAvailabilityReportOptions` here in that case?

aaemnnosttv reviewed (changes requested):
Thanks @nfmohit – a few points to address for you.

One thing that I noticed is that we don't seem to be handling a view-only context in `withCustomDimensions` where all the syncing and dimension creation shouldn't be attempted. I could have missed it and it's not the focus of this issue but something we want to make sure is covered for launch.

nfmohit commented on assets/js/modules/analytics-4/components/widgets/PopularAuthorsWidget.stories.js:
```diff
@@ -204,11 +206,17 @@ export default {
 				} );
+				registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetProperty(
+					{
+						createTime: '2014-10-02T15:01:23Z',
```
This just needs to be a valid date that is well behind for our GA4 property to not be in a gathering data state. I had actually copied this over from the GA4 fixtures. See:

https://github.com/google/site-kit-wp/blob/14d70dea611e7de142ce3f6c122f9be15bbcae5a/assets/js/modules/analytics-4/datastore/__fixtures__/properties.json#L6

Do we need to change it?

aaemnnosttv replied:
Ah that's fine then, I was mostly curious since it seems arbitrary. If the date has some significance, it would be good to make that more explicit so that it's less likely to be broken unintentionally.

nfmohit commented:
> Thanks @nfmohit – a few points to address for you.
> 
> One thing that I noticed is that we don't seem to be handling a view-only context in `withCustomDimensions` where all the syncing and dimension creation shouldn't be attempted. I could have missed it and it's not the focus of this issue but something we want to make sure is covered for launch.

Thank you for the kind and detailed review, @aaemnnosttv. I've addressed most of the feedback and left follow-up questions to some. Please let me know how it looks, thanks!

aaemnnosttv reviewed (approved):
LGTM, thanks @nfmohit !
